### DISCUSSION TITLE
Discovery protocol ports

### DISCUSSION THREAD

**Discussion by ChinmayGameti** [eProsima/Fast-DDS#3647] [Q&A]
Hello, 
I have a doubt regarding the Discovery protocol at the OS-abstraction level.
Discovery protocols use some specific ports for the discovery of other publishers or subscribers in the system(using some equation given in OMG DDS standard), and it is only meant to be consumed by the DDS layer. Are these ports accessible to ROS applications are the program level, meaning can a process sniff packets using this port or send packets using this discovery port?

**Comment by JLBuenoLopez** [comment]
Hi @ChinmayGameti,

You can find information related to this topic in the [Listening Locators section](https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/listening_locators.html#listening-locators) in Fast DDS documentation. In ROS 2 you can configure the listening locators using a [XML configuration file](https://fast-dds.docs.eprosima.com/en/latest/fastdds/ros2/ros2_configure.html#xml-configuration). ROS 2 does not expose Fast DDS API which can give you the desired information: [`DataWriter::get_sending_locators`](https://fast-dds.docs.eprosima.com/en/latest/fastdds/api_reference/dds_pim/publisher/datawriter.html#_CPPv4NK8eprosima7fastdds3dds10DataWriter20get_sending_locatorsERN4rtps11LocatorListE) and [`DataReader::get_listening_locators`](https://fast-dds.docs.eprosima.com/en/latest/fastdds/api_reference/dds_pim/subscriber/datareader.html#_CPPv4NK8eprosima7fastdds3dds10DataReader22get_listening_locatorsERN4rtps11LocatorListE).

Edit: I have been looking into ROS 2 Humble API and the previous methods are exposed in `rcl_subscription_get_network_flow_endpoints` and `rcl_publisher_get_network_flow_endpoints`. Also, the native Fast DDS DataWriter is accessible in the RMW layer through the method `rmw_fastrtps_cpp::get_datawriter` and similarly the DataReader with `rmw_fastrtps_cpp::get_datareader` method.
